src/alphafold_impact/pipelines/data_processing_nih_clinical_trials/nodes.py

```diff
@@ -0,0 +1,147 @@
+"""
+This module contains the functions used in the data processing pipeline for
+the NIH clinical trials data.
+
+Functions:
+    trials_with_references: Loads NIH clinical trials from partitioned
+        dataset, filters out trials that do not reference a paper,
+        concatenates results into one DataFrame.
+    nih_clinical_trials_links_to_papers: Produces a DataFrame which can
+        be used to link NIH clinical trials to research papers via 
+        PubMed ID or DOI.
+"""
+from typing import Dict, Callable, Any
+import logging
+import ast
+import re
+import pandas as pd
+
+
+logger = logging.getLogger(__name__)
+
+
+def trials_with_references(
+    nih_clinical_trials_partitions: Dict[str, Callable[[], Any]]
+) -> pd.DataFrame:
+    """Loads NIH clinical trials from partitioned dataset, filters out trials
+    that do not reference a paper, concatenates results into one DataFrame.
+
+    Args:
+        nih_clinical_trials_partitions (dict): Partitioned dataset of NIH clinical trials.
+
+    Returns:
+        DataFrame of NIH clinical trials with a paper reference.
+    """
+    ct_with_refs_combined = pd.DataFrame()
+    n_partitions = len(nih_clinical_trials_partitions)
+
+    for idx, (_, ct_partition_load_func) in enumerate(
+        sorted(nih_clinical_trials_partitions.items()), start=1
+    ):
+        ct_partition_data_with_refs = ct_partition_load_func().query(
+            "`ProtocolSection.ReferencesModule.ReferenceList.Reference`.notna()"
+        )
+
+        ct_with_refs_combined = pd.concat(
+            [ct_with_refs_combined, ct_partition_data_with_refs],
+            ignore_index=True,
+            join="outer",
+        )
+
+        logger.info(
+            "Concatenated clinical trials with references from partition %s/%s",
+            idx,
+            n_partitions,
+        )
+
+    return ct_with_refs_combined
+
+
+def _reference_string_to_df(row: pd.Series) -> pd.DataFrame:
```

Comment:
Neat! I wonder whether we should agree on how inner use functions are used - I may have a preference for listing them all at the top, but feel free to disregard.

Reply:
I'm happy to have them at the top. I've moved them